packages/request/lib/Agent.js

```diff
@@ -0,0 +1,341 @@
+//
+// SPDX-FileCopyrightText: 2020 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+'use strict'
+
+const http2 = require('http2')
+const net = require('net')
+const crypto = require('crypto')
+const { promisify } = require('util')
+const QuickLRU = require('quick-lru')
+const { globalLogger: logger } = require('@gardener-dashboard/logger')
+const Semaphore = require('./Semaphore')
+
+const kSemaphore = Symbol('semaphore')
+const kRequest = Symbol('request')
+const kSid = Symbol('sid')
+const kTimestamp = Symbol('timestamp')
+const kConnected = Symbol('connected')
+const kTimeoutId = Symbol('timeoutId')
+const kIntervalId = Symbol('intervalId')
+
+const { NGHTTP2_CANCEL } = http2.constants
+
+class Agent {
+  constructor (options = {}) {
+    const {
+      maxCachedTlsSessions = 100,
+      timeout = 60000,
+      connectTimeout = 5000,
+      gracePeriod = 5000,
+      maxOutstandingPings = 2,
+      pingInterval = 30000
+    } = options
+    this.timeout = timeout
+    this.connectTimeout = connectTimeout
+    this.gracePeriod = gracePeriod
+    this.maxOutstandingPings = maxOutstandingPings
+    this.pingInterval = pingInterval
+    this.tlsSessionCache = new QuickLRU({
+      maxSize: maxCachedTlsSessions
+    })
+    this.sessionMap = new Map()
+  }
+
+  destroy () {
+    for (const iterable of this.sessionMap.values()) {
+      for (const session of iterable) {
+        this.clearSessionTimers(session)
+        this.deleteSession(session, { force: true })
+        session.destroy()
+        this.tlsSessionCache.delete(session[kSid])
+      }
+    }
+  }
+
+  findSession (sid) {
+    const iterable = this.sessionMap.get(sid)
+    if (iterable) {
+      return Array.from(iterable)
+        // consider free sessions only
+        .filter(({ [kSemaphore]: semaphore }) => {
+          return semaphore.value > 0
+        })
+        // session with the highest load first
+        .sort(({ [kSemaphore]: a }, { [kSemaphore]: b }) => {
+          return a.value - b.value
+        })
+        .shift()
+    }
+  }
+
+  getSession (authority, { settings, ...options } = {}) {
```

Comment:
Add docu or example in PR how to use the Agent. You should only get a session instance if you synchronously do a request.

Reply:
Request method has been added to the Agent. The client does not know about sessions anymore. This is handled internally by the Agent. Examples to the docu will be added in the future.